Synchronized, Volatile and Blocking queue

Yanetta opened the pull request.

Yanetta committed:
1commit on branch3

 2 commit on branch3

SerhiiPasko commented on src/main/java/multuthreadingAndBlockingQueue/Account.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    public void withdrawDollars(long amount) {
+        synchronized (this) {
```
What this give you?

Yanetta replied:
This synchronized block works like synchronized method. I have two threads which work with  withdrawDollars method. WithdrawDollars method changes dollarBalance variable on my single object "account".  I synchronized this method on this object . Thank to this only one thread will have a possibility to execute code in a synchronized block at certain moment of time.

SerhiiPasko commented on src/main/java/multuthreadingAndBlockingQueue/Account.java:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    public void depositHryvhya(long amount) {
```
I think, you wrote same logic for gruvna and dollar for show synchonised only?

Yanetta replied:
Methods for hryvnya are not synchronized, but work with volatile variable hryvnyaBalance;

SerhiiPasko commented on src/main/java/multuthreadingAndBlockingQueue/Producer.java:
```diff
@@ -0,0 +1,28 @@
+    private BlockingQueue <Integer> bq;
+
+    public Producer(BlockingQueue <Integer> queue) {
+        this.bq = queue;
```
What reason of "this" in this case?

Yanetta replied:
Just to emphasize that it is the variable of this object )). I will remove "this".

SerhiiPasko commented on src/main/java/multuthreadingAndBlockingQueue/Main.java:
```diff
@@ -0,0 +1,91 @@
+        }
+
+        System.out.println("BlockingQueue examples");
+        BlockingQueue<Integer> bq = new LinkedBlockingQueue<>(1);
```
For one element present special collections

Yanetta replied:
OK, I'll make changes. I just wanted to examine how method take removes elements)

SerhiiPasko reviewed (approved):
Good

SerhiiPasko reviewed (changes requested):
Sorry, for previous result. I don't see synchronized block

Yanetta commented:
My synchronized block is inside withdrawDollars method, or it is not what you mean?

Yanetta committed:
Update Main.java

Capacity of BlockingQueue bq changed from 1 to 10.

Yanetta committed:
Update Producer.java

Removed "this" in constructor

Yanetta committed:
Update Main.java

Classes InterestPayment and InterestInfo were added. Their methods are synchronized on account object.

SerhiiPasko commented on src/main/java/multuthreadingAndBlockingQueue/Main.java:
```diff
@@ -0,0 +1,142 @@
+        new Thread(interestInfo).start();
+    }
+}
+class InterestPayment implements Runnable {
```
Why did you do same logic in main? SOLID, "S"

SerhiiPasko reviewed (approved):
Good. Read comments
